### PR TITLE
[MPS] Add Unique and unique_consecutive ops.

### LINKED ISSUE #88487: [MPS] Add support for aten::unique_consecutive for MPS backend
### 🐛 Describe the bug

```python
classifier = pipeline("zero-shot-classification",
                      model="facebook/bart-large-mnli", device = torch.device("mps"))
candidate_labels = ['label1', 'label2', 'label3'] #list of labels to tag
classifier("SAMPLE TEXT INPUT" , candidate_labels)
```
```
NotImplementedError: The operator 'aten::unique_consecutive' is not current implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.
```

### Versions

NA

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by kulinseth**
Add check for macos 13.0

Fixes #88487

### CODE REVIEW

**Comment by kulinseth** [comment]
@malfet , rebased . Please give it a review.

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+                                                               secondaryTensor:frontNMinusOne
+                                                                          name:nil];
+  MPSGraphTensor *mask = [graph castTensor:notEqualToPreviousElement
+                                    toType:MPSDataTypeInt32
+                                      name:@"castMaskTensor"];
+
+  // If comparing tensors, not scalars, check if entire tensor matches previos element using reductionOr over tensor
+  if (dimOpt.has_value() && [shape count] != 1) {
+    NSMutableArray *axes = [[NSMutableArray alloc] initWithCapacity:[shape count]-1];
+    for (NSUInteger axis = 0; axis < [shape count]; axis++){
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Unique.mm:108]:
```suggestion
    for (NSUInteger axis = 0; axis < [shape count]; axis++) {
```

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+      if (axis != dim)
+        [axes addObject:[NSNumber numberWithUnsignedInteger:axis]];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Unique.mm:110]:
```suggestion
      if (axis != dim) {
        [axes addObject:[NSNumber numberWithUnsignedInteger:axis]];
      }  
```

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+                                            indicesTensor:maskedIndicesWithHead
+                                                    shape:destShape
+                                                     axis:dim
+                                                     mode:MPSGraphScatterModeSet
+                                                     name:nil];
+  // Cast back if necessary
+  if ([uniqueGraph->inputTensor_ dataType] != dataType)
+    resultTensor = [graph castTensor:resultTensor
+                              toType:[uniqueGraph->inputTensor_ dataType]
+                                name:@"castResultTensor"];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Unique.mm:157]:
```suggestion
  if ([uniqueGraph->inputTensor_ dataType] != dataType) {
    resultTensor = [graph castTensor:resultTensor
                              toType:[uniqueGraph->inputTensor_ dataType]
                                name:@"castResultTensor"];
  }
```

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+  return std::make_tuple(output, inverse_indices, counts);
+}
+
+std::tuple<Tensor, Tensor, Tensor>
+unique_consecutive_mps(const Tensor& self, const bool return_inverse, const bool return_counts, c10::optional<int64_t> dim) {
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("MPS: unique_consecutive op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+    return at::unique_consecutive(self.to("cpu"), return_inverse, return_counts, dim);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Unique.mm:316]: Don't you need to copy result tensors back to MPS?

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+  return _unique_impl_mps(self, return_inverse, return_counts, true, dim);
+}
+
+std::tuple<Tensor, Tensor, Tensor>
+unique_dim_consecutive_mps(const Tensor& self, int64_t dim, const bool return_inverse, const bool return_counts) {
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("MPS: unique_dim_consecutive op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+    return at::unique_dim_consecutive(self.to("cpu"), dim, return_inverse, return_counts);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Unique.mm:327]: Same as above, result tensors must be copied back to MPS

**Review by malfet — APPROVED** [review]
Looks good to me. 
Major areas of concern:
- Large tensors (add overflow checks when computing number of elements)
- For the CPU fallbacks, aren't results needed to be copied back to MPS?

Nits:
 - Formatting: `if ()` block should always be surrounded by `{` and don't one needs spaces after objc argument name?
 - std::array vs NSArray (this way one can avoid `[NSNull null]`)

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+           // Initialize graph
+           MPSGraph* mpsGraph = make_mps_graph();
+           newCachedGraph = new UniqueCachedGraph(mpsGraph);
+
+           // Workaround for MPSShaderLibrary bug
+           // TODO: Remove once https://github.com/pytorch/pytorch/issues/82305 is resolved
+           auto inputType = getMPSScalarType(self.scalar_type());
+           newCachedGraph->inputTensor_ = mpsGraphRankedPlaceHolder(mpsGraph, inputType, getMPSShape(self.sizes()));
+
+           NSArray<MPSGraphTensor *> *outputTensors = buildUniqueGraph(self, newCachedGraph, return_inverse, return_counts, consecutive, dim);
```
[the remark this answers is not in the document]
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Thanks - I've updated it

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+  MPSGraphTensor *inputTensor = uniqueGraph->inputTensor_;
+  MPSShape *shape = [inputTensor shape];
+  MPSShape *destShape = shape;
+  NSUInteger length = [shape[dim] integerValue];
+  MPSDataType dataType = [inputTensor dataType];
+
+  MPSGraphTensor *resultTensor = (MPSGraphTensor *)[NSNull null];
+  MPSGraphTensor *inverseIndicesTensor = (MPSGraphTensor *)[NSNull null];
+  MPSGraphTensor *countTensor = (MPSGraphTensor *)[NSNull null];
+  MPSGraphTensor *lengthTensor = (MPSGraphTensor *)[NSNull null];
```
[the remark this answers is not in the document]
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Thanks
Updated

File: aten/src/ATen/native/mps/operations/Unique.mm (hunk 2 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

File: aten/src/ATen/native/mps/operations/Unique.mm (hunk 3 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

File: aten/src/ATen/native/mps/operations/Unique.mm (hunk 1 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

File: aten/src/ATen/native/mps/operations/Unique.mm (hunk 4 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Thanks - good catch. 
I've updated it.

File: aten/src/ATen/native/mps/operations/Unique.mm (hunk 5 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+                                const bool return_inverse, const bool return_counts,
+                                const bool consecutive, c10::optional<int64_t> dimOpt)
+{
+  return "_unique2_mps:" + getMPSTypeString(dtype) + "[" + getArrayRefString(base_shape) +
+         "]:[" + (dimOpt.has_value() ? to_string(dimOpt.value()) : "None") + "]:[" + to_string(return_inverse) +
+         "]:[" + to_string(return_counts) + "]:[" + to_string(consecutive) + "]";
+}
+
+// dim arg not supported when non consecutive, ie sorted
+NSArray<MPSGraphTensor*> *buildUniqueGraph(const Tensor& self, UniqueCachedGraph *uniqueGraph, const bool return_inverse, const bool return_counts, const bool consecutive, c10::optional<int64_t> dimOpt) {
```
[the remark this answers is not in the document]
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Thanks. Updated

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+  int64_t dim = dimOpt.has_value() ? maybe_wrap_dim(dimOpt.value(), self.dim()) : 0;
+
+  MPSGraph *graph = uniqueGraph->graph();
+  MPSGraphTensor *inputTensor = uniqueGraph->inputTensor_;
+  MPSShape *shape = [inputTensor shape];
+  MPSShape *destShape = shape;
+  NSUInteger length = [shape[dim] integerValue];
```
[the remark this answers is not in the document]
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

File: aten/src/ATen/native/mps/operations/Unique.mm [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+  }
+
+  bool needsFlatten = !(dimOpt.has_value() || [shape count] == 1);
+  if (needsFlatten) {
+    inputTensor = [graph reshapeTensor:inputTensor
+                             withShape:@[@-1]
+                                  name:nil];
+    length = 1;
+    for(NSUInteger i = 0; i < [shape count]; i++)
+      length *= [shape[i] integerValue];
```
[the remark this answers is not in the document]
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Unique.mm]: Updated

**Comment by jkoudys** [comment]
I'm building this branch myself, and I get this warning an awful lot:
```
clang: warning: argument unused during compilation: '-mfpu=neon' [-Wunused-command-line-argument]
```
on my macos 13.2 M1. I'm building the usual recommended pytorch way:
```
CC=clang CXX=clang++ python3 setup.py develop
```

**Comment by DenisVieriu97** [comment]
@jkoudys this change should get into the next pytorch nightly binary - could you please try that and let me know if you see any issues?

**Comment by tj-safely** [comment]
Is this merged on master? It looks like it is, but I am unable to use the op in yolov5 and pytorch 1.13.1, and I have tested on the nightly too.

**Comment by DenisVieriu97** [comment]
@tj-safely nightly builds should have them. Can you try doing a force reinstall to make sure it got installed properly? `pip3 install --pre --force-reinstall torch --index-url https://download.pytorch.org/whl/nightly/cpu`

**Comment by tj-safely** [comment]
@DenisVieriu97  I have tried a few different methods for installing a matching torchvision package to no success.

Right now I have this:
torch                   2.0.0.dev20230220
torchaudio              2.0.0.dev20230221
torchvision             0.15.0a0+a46d97c

And get this error message:
  F```
ile "/Users/tj/yolov5/utils/general.py", line 975, in non_max_suppression
    i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
  File "/Users/tj/opt/anaconda3/envs/yolo_t3/lib/python3.10/site-packages/torchvision/ops/boxes.py", line 40, in nms
    _assert_has_ops()
  File "/Users/tj/opt/anaconda3/envs/yolo_t3/lib/python3.10/site-packages/torchvision/extension.py", line 48, in _assert_has_ops
    raise RuntimeError(
RuntimeError: Couldn't load custom C++ ops. This can happen if your PyTorch and torchvision versions are incompatible, or if you had errors while compiling torchvision from source. For further information on the compatible versions, check https://github.com/pytorch/vision#installation for the compatibility matrix. Please check your PyTorch version with torch.
```

When I add torchvision to your command or try with only torchvision, it attempts to install many different versions of torchvision and does not install any of them.
It says they depend on numpy and pillow which I have:
numpy==1.24.2
pillow==9.3.0

Finally, I have this torchvision version installed by running this command:
pip install git+https://github.com/pytorch/vision.git

which I found from this link:
https://github.com/pytorch/vision/issues/833

Thank you for your help and further guidance is appreciated.

**Comment by DenisVieriu97** [comment]
@tj-safely try uninstalling all previous packages of torch/torchvision/torchaudio, then do a clean install. 
E.g execute the following in a terminal (in your environment):
```
for i in {1..20}; do pip uninstall -y torch torchvision torchaudio; done
```
then:
```
pip3 install --pre --force-reinstall torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu
```

**Comment by tj-safely** [comment]
I am still unable to install torchvision this way. To reproduce:
` conda create -n yolo python`
`conda activate yolo`
```
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
`pip3 install -r requirements.txt`
`for i in {1..20}; do pip uninstall -y torch torchvision torchaudio; done`
`pip3 install --pre --force-reinstall torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu`

Everything works successfully until the last command:
```
Looking in indexes: https://download.pytorch.org/whl/nightly/cpu
Collecting torch
  Using cached https://download.pytorch.org/whl/nightly/cpu/torch-2.0.0.dev20230220-cp310-none-macosx_10_9_x86_64.whl (140.0 MB)
Collecting torchvision
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230223-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
Collecting torchaudio
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchaudio-2.0.0.dev20230223-cp310-cp310-macosx_10_9_x86_64.whl (3.9 MB)
Collecting networkx
  Using cached https://download.pytorch.org/whl/nightly/networkx-3.0rc1-py3-none-any.whl (2.0 MB)
Collecting sympy
  Using cached https://download.pytorch.org/whl/nightly/sympy-1.11.1-py3-none-any.whl (6.5 MB)
Collecting typing-extensions
  Using cached https://download.pytorch.org/whl/nightly/typing_extensions-4.4.0-py3-none-any.whl (26 kB)
Collecting filelock
  Using cached https://download.pytorch.org/whl/nightly/filelock-3.9.0-py3-none-any.whl (9.7 kB)
Collecting requests
  Using cached https://download.pytorch.org/whl/nightly/requests-2.28.1-py3-none-any.whl (62 kB)
Collecting torchvision
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230222-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230221-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230220-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230219-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230218-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230217-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230216-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230215-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230214-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230213-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230212-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230211-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230210-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230209-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230208-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230207-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230206-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230205-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230204-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230203-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230202-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230201-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230131-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230130-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230129-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230128-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230127-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230126-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230125-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230124-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230123-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230122-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230121-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230120-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230119-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230118-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230117-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230116-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230115-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230114-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230113-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230112-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230111-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230110-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230109-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230108-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230107-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230106-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230105-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230104-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230103-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230102-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20230101-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/cpu/torchvision-0.15.0.dev20221231-cp310-cp310-macosx_10_9_x86_64.whl (1.6 MB)
  Using cached https://download.pytorch.org/whl/nightly/torchvision-0.14.0.dev20220816-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
  Using cached https://download.pytorch.org/whl/nightly/torchvision-0.14.0.dev20220815-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
  Using cached https://download.pytorch.org/whl/nightly/torchvision-0.14.0.dev20220814-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
  Using cached https://download.pytorch.org/whl/nightly/torchvision-0.14.0.dev20220813-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
  Using cached https://download.pytorch.org/whl/nightly/torchvision-0.14.0.dev20220812-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.
Collecting torch
  Using cached https://download.pytorch.org/whl/nightly/cpu/torch-2.0.0.dev20230219-cp310-none-macosx_10_9_x86_64.whl (140.0 MB)
ERROR: Cannot install torchvision==0.14.0.dev20220812, torchvision==0.14.0.dev20220813, torchvision==0.14.0.dev20220814, torchvision==0.14.0.dev20220815, torchvision==0.14.0.dev20220816, torchvision==0.15.0.dev20221231, torchvision==0.15.0.dev20230101, torchvision==0.15.0.dev20230102, torchvision==0.15.0.dev20230103, torchvision==0.15.0.dev20230104, torchvision==0.15.0.dev20230105, torchvision==0.15.0.dev20230106, torchvision==0.15.0.dev20230107, torchvision==0.15.0.dev20230108, torchvision==0.15.0.dev20230109, torchvision==0.15.0.dev20230110, torchvision==0.15.0.dev20230111, torchvision==0.15.0.dev20230112, torchvision==0.15.0.dev20230113, torchvision==0.15.0.dev20230114, torchvision==0.15.0.dev20230115, torchvision==0.15.0.dev20230116, torchvision==0.15.0.dev20230117, torchvision==0.15.0.dev20230118, torchvision==0.15.0.dev20230119, torchvision==0.15.0.dev20230120, torchvision==0.15.0.dev20230121, torchvision==0.15.0.dev20230122, torchvision==0.15.0.dev20230123, torchvision==0.15.0.dev20230124, torchvision==0.15.0.dev20230125, torchvision==0.15.0.dev20230126, torchvision==0.15.0.dev20230127, torchvision==0.15.0.dev20230128, torchvision==0.15.0.dev20230129, torchvision==0.15.0.dev20230130, torchvision==0.15.0.dev20230131, torchvision==0.15.0.dev20230201, torchvision==0.15.0.dev20230202, torchvision==0.15.0.dev20230203, torchvision==0.15.0.dev20230204, torchvision==0.15.0.dev20230205, torchvision==0.15.0.dev20230206, torchvision==0.15.0.dev20230207, torchvision==0.15.0.dev20230208, torchvision==0.15.0.dev20230209, torchvision==0.15.0.dev20230210, torchvision==0.15.0.dev20230211, torchvision==0.15.0.dev20230212, torchvision==0.15.0.dev20230213, torchvision==0.15.0.dev20230214, torchvision==0.15.0.dev20230215, torchvision==0.15.0.dev20230216, torchvision==0.15.0.dev20230217, torchvision==0.15.0.dev20230218, torchvision==0.15.0.dev20230219, torchvision==0.15.0.dev20230220, torchvision==0.15.0.dev20230221, torchvision==0.15.0.dev20230222 and torchvision==0.15.0.dev20230223 because these package versions have conflicting dependencies.

The conflict is caused by:
    torchvision 0.15.0.dev20230223 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230222 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230221 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230220 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230219 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230218 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230217 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230216 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230215 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230214 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230213 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230212 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230211 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230210 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230209 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230208 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230207 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230206 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230205 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230204 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230203 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230202 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230201 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230131 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230130 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230129 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230128 depends on pillow!=8.3.* and >=5.3.0
    torchvision 0.15.0.dev20230127 depends on numpy
    torchvision 0.15.0.dev20230126 depends on numpy
    torchvision 0.15.0.dev20230125 depends on numpy
    torchvision 0.15.0.dev20230124 depends on numpy
    torchvision 0.15.0.dev20230123 depends on numpy
    torchvision 0.15.0.dev20230122 depends on numpy
    torchvision 0.15.0.dev20230121 depends on numpy
    torchvision 0.15.0.dev20230120 depends on numpy
    torchvision 0.15.0.dev20230119 depends on numpy
    torchvision 0.15.0.dev20230118 depends on numpy
    torchvision 0.15.0.dev20230117 depends on numpy
    torchvision 0.15.0.dev20230116 depends on numpy
    torchvision 0.15.0.dev20230115 depends on numpy
    torchvision 0.15.0.dev20230114 depends on numpy
    torchvision 0.15.0.dev20230113 depends on numpy
    torchvision 0.15.0.dev20230112 depends on numpy
    torchvision 0.15.0.dev20230111 depends on numpy
    torchvision 0.15.0.dev20230110 depends on numpy
    torchvision 0.15.0.dev20230109 depends on numpy
    torchvision 0.15.0.dev20230108 depends on numpy
    torchvision 0.15.0.dev20230107 depends on numpy
    torchvision 0.15.0.dev20230106 depends on numpy
    torchvision 0.15.0.dev20230105 depends on numpy
    torchvision 0.15.0.dev20230104 depends on numpy
    torchvision 0.15.0.dev20230103 depends on numpy
    torchvision 0.15.0.dev20230102 depends on numpy
    torchvision 0.15.0.dev20230101 depends on numpy
    torchvision 0.15.0.dev20221231 depends on numpy
    torchvision 0.14.0.dev20220816 depends on numpy
    torchvision 0.14.0.dev20220815 depends on numpy
    torchvision 0.14.0.dev20220814 depends on numpy
    torchvision 0.14.0.dev20220813 depends on numpy
    torchvision 0.14.0.dev20220812 depends on numpy

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

**Comment by DenisVieriu97** [comment]
@tj-safely could you please try with conda instead pip?
```conda install --force-reinstall pytorch torchvision torchaudio -c pytorch-nightly```

**Pull request closed without merging**